### PR TITLE
[Merged by Bors] - feat: factor out backtracking code from solve_by_elim

### PR DESCRIPTION
**Pull request by kim-em**
Following up from discussion on [zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/.60mono.60.20changes.3F).

No change in functionality, this is just a refactor.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/SolveByElim.lean [hunk 1]
```diff
@@ -99,6 +73,114 @@ structure Config extends ApplyConfig where
   If failure, we backtrack. (defaults to failure) -/
   discharge : MVarId → MetaM (Option (List MVarId)) := fun _ => failure
 
+
+def backtracking (cfg : Config) (alternatives : MVarId → MetaM (List (MetaM (List MVarId))))
```
Comment (thorimur) [inline Mathlib/Tactic/SolveByElim.lean:77]: Could `backtracking` be given a `trace` argument like `applyTactics` for use as a component in other tactics? (Or is it meant just for `solve_by_elim`? If so, should it be in the `SolveByElim` namespace?)

File: Mathlib/Tactic/SolveByElim.lean [hunk 2]
```diff
@@ -182,6 +268,12 @@ def elabContextLemmas (g : MVarId) (lemmas : List (TermElabM Expr)) (ctx : TermE
     MetaM (List Expr) := do
   g.withContext (Elab.Term.TermElabM.run' do pure ((← lemmas.mapM id) ++ (← ctx)))
 
+/-- Returns the list of tactics corresponding to apply the available lemmas to the goal. -/
```
Comment (thorimur) [inline Mathlib/Tactic/SolveByElim.lean:271]: Missing word?

File: Mathlib/Tactic/Nondeterministic.lean [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+if they preclude finding solutions to later goals).
+-/
+
+/--
+`nondeterministicMany` runs a "nondeterministic" tactic
+`tac : MVarId → MetaM (List (MetaM (List MVarId)))`
+recursively on a collection of goals, backtracking as necessary.
+For each subgoal, it evaluates `recurse : MVarId → MetaM Bool` to decide whether
+to continue solving that goal, or to "suspend" the goal, returning it to the caller. -/
+def Lean.MVarId.nondeterministicMany
```
[the remark this answers is not in the document]
Comment (semorrison) [reply Mathlib/Tactic/Nondeterministic.lean:36]: I'm going to rip this out of this PR, for now. I have plans for further refactoring in terms of tools for handling nondeterministic tactics, but I don't want to do it here.

File: Mathlib/Tactic/Backtracking.lean [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+      if (← cfg.suspend g) then
+        withTraceNode trace
+          (fun _ => return m!"⏸️ suspending search and returning as subgoal") do
+        run (n+1) gs (g :: acc)
+      else
+        try
+          -- We attempt to find an expression which can be applied,
+          -- and for which all resulting sub-goals can be discharged using `run n`.
+          g.firstContinuation alternatives (fun res => run n (res ++ gs) acc)
+        catch _ =>
```
Comment (eric-wieser) [inline Mathlib/Tactic/Backtracking.lean:148]: Should we trace the error in the `_`?

**Review by kmill — APPROVED** [review]
The only thing I might have suggested is that `acc` could be an `Array`, but that's not a good idea since the program is actually taking advantage of sharing (from the `g.firstContinuation` line).

The backtracking API seems reasonable enough and it doesn't seem like you're really jumping through any hoops to use it, so I approve.

bors d+

File: Mathlib/Tactic/Backtracking.lean (hunk 4 above)
Comment (semorrison) [reply Mathlib/Tactic/Backtracking.lean:148]: There's no error here, it is just `firstContinuation` failing to find anything that works.

**Comment by semorrison** [comment]
bors merge

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**